File: client/src/lib/models/routine.ts
import type { Workout } from "./workout";

export class Routine {
    id: string;
    user_id: string;
    name: string;
    days: string[];
    last_completed: string;
    note: string;
    workouts: Workout[];
}

export class RoutineRow {
    id: string;
    user_id: string;
    name: string;
    days: string[];
    last_completed: string;
    note: string;
    workouts: string[];
}

export class InsertRoutineRow {
    user_id: string;
    name: string;
    days: string[];
    last_completed: string;
    note: string;
    workouts: string[];
}

export class RoutineHistoryRow {
    id: string;
    routine: string;
    user_id: string;
    completed_on: string;
    duration_in_sec: number;
    workouts: RoutineHistoryWorkout[];
}

export class RoutineHistory {
    id: string;
    routine: Routine;
    user_id: string;
    completed_on: Date;
    duration_in_sec: number;
    workouts: RoutineHistoryWorkout[];
}

export class RoutineHistoryWorkout {
    name: string;
    weight: number;
    reps: number;
}

File: server/src/model/routine.rs
use super::{
    shared_types::{RoutineHistoryRef, RoutineRef, UserRef, WorkoutRef, Set},
    workout::Workout,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize)]
pub struct Routine {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<RoutineRef>,
    pub user_id: UserRef,
    pub name: String,
    pub days: Vec<String>,
    pub last_completed: String,
    pub note: String,
    pub workouts: Vec<Workout>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct RoutineRow {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<RoutineRef>,
    pub user_id: UserRef,
    pub name: String,
    pub days: Vec<String>,
    pub last_completed: String,
    pub note: String,
    pub workouts: Vec<WorkoutRef>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct RoutineHistoryRow {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<RoutineHistoryRef>,
    pub user_id: UserRef,
    pub routine_id: RoutineRef,
    pub name: String,
    pub completed_on: String,
    pub duration_in_sec: i32,
    pub workouts: Vec<RoutineHistoryWorkout>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct RoutineHistoryWorkout {
    pub name: String,
    pub sets: Vec<Set>,
}